### ISSUE TITLE
Add `srcpaths` setting

### ISSUE THREAD

**Issue by okken** [pytest-dev/pytest#9114]
#### What's the problem this feature will solve?

pytest seems designed to test Python packages.
However, it's getting used to test applications as well. 
The main problem with testing applications is telling pytest how to find the code under test. 
Specifically, `sys.path` needs modified to include the directory where the code under test lives.

#### Describe the solution you'd like

I'd like a `srcpaths` setting. We already have `testpaths`. 
`srcpaths` would be a list of directories to add to `sys.path`.

Any project that isn't an installable package (installable through pip) that wants the source code and test code in different directories needs this.

#### Alternative Solutions

* Roll your own solutions
* Plugins 
    * [pytest-srcpaths](https://github.com/okken/pytest-srcpaths) does just this thing.
    * [pytest-pythonpaths](https://github.com/ericpalakovichcarr/pytest-pythonpath) does more
        * does this
        * also has a site_dirs feature which I still don't understand
        * also (undocumented) adds the current directory to sys.path

**Comment by okken** [comment]
I wouldn't mind working on this. It shouldn't be too much code.  
But...
1. I'm not sure where to put the code. Any ideas? It could be as a builtin plugin.
2. Is this something the core team would be ok with having in core? Or do you feel it should stay in plugin land?

**Comment by RonnyPfannschmidt** [comment]
an non-package application would not use a src layout
thus having a normal tests folder with a `__init__.py` would take care of all the paths

can you elaborate on setups where there is a need for more?

**Comment by okken** [comment]
I think "a non-package application would not use a src layout" is too strong of a statement.  

1. I prefer to use src layout for many reasons, not just for testing.  
I like the separation of top level directories into src, docs, tests.

2. The `__init__.py` solution assumes that importing from the top level is convenient.
I could see having a layer deep within an application that I'd like to write tests against without having to specify the location within the application to import it from for testing.

3. There are other uses of pytest where the lack of this feature is a problem, otherwise the plugins mentioned above wouldn't be downloaded at all. But they are: 
    * https://pypistats.org/packages/pytest-pythonpath  
    * https://pypistats.org/packages/pytest-srcpaths
  
4. The use of `__init__.py` within tests directory continues to confuse people. We used to tell people not to put it at the top tests directory. Now the docs say that's fine for application usage, if people get that far into the docs, after we've told them explicitly to use src layout.

5. Some applications start out as non-packages and then turn into packages. Telling people they have to modify their layout to accommodate testing is weird.

So, those aren't examples. Just reasons.

**Comment by bluetech** [comment]
I have a project where I use pytest-pythonpath, would love to see it integrated. My layout is like

```
library1/
library2/
django_project_1/
django_project_2/
integration_tests/
```

and I use `python_paths = ..` in the `django_project{1,2}/pytest.ini` so they can reference library1/2, and also `integration_tests` does `python_paths = .. django_project1`.

Over the years I tried different layouts but this works best.

Now the plugin is very simple, so I tried to just inline it into the repo to avoid the dependency. But it's hard (impossible?) to do this in a plugin from within the repo itself -- it needs to run as early as possible. Also, I would have to duplicate it in multiple sub-directories.

Regarding `pytest-srcpaths`, some comments:

- The name `pythonpath` seems more discoverable to me than `srcpaths`, ditto for the pytest setting.
- `pytest-pythonpath` inserts the user-given paths in `reversed()` -- so if `python_paths = a b` then `sys.path` would be `[a, b, ...]` -- seems more intuitive than `[b, a, ...]`.
- `pytest-pythonpath` uses `pytest_load_initial_conftests` with `tryfirst`, which runs earlier to `pytest_configure` -- probably better?
- I also don't know what's the `site_dirs` stuff that `pytest-pythonpath` does.
- I don't know why `pytest-pythonpath` inserts the cwd to the sys.path.
